### PR TITLE
Added undo / redo stack for cropper

### PR DESCRIPTION
**Pull request by lescobar723**
I added a stack to keep track of each cropper state whenever the cropper is adjusted or edited. The user can now press the 'z' key for undo or 'y' key for redo and the cropper will be adjusted accordingly. 

This added feature is in response to #976

### CODE REVIEW

**Comment by sindresorhus** [comment]
It should use Shift+Command+Z to redo. That's the convention on macOS.

File: renderer/containers/cropper.js [hunk 1]
```diff
@@ -334,6 +356,50 @@ export default class CropperContainer extends Container {
...
+
+  undo = () => {
+    if (this.state.cropperPointer > 0) {
+      const originalCropperHistory = this.state.cropperHistory;
+      const currPointer = this.state.cropperPointer - 1;
+      const newState = this.state.cropperHistory[currPointer];
+      this.setState(newState);
+      this.setState({undone: true});
+      this.setState({cropperHistory: originalCropperHistory});
+      this.setState({cropperPointer: currPointer});
```
Comment (sindresorhus) [inline renderer/containers/cropper.js:389]: I think it would be better to use a single setState call.

File: renderer/containers/cropper.js [hunk 2]
```diff
@@ -334,6 +356,50 @@ export default class CropperContainer extends Container {
...
+        width,
+        height,
+        ratio,
+        displayId
+      });
+    }
+  }
+
+  undo = () => {
+    if (this.state.cropperPointer > 0) {
```
Comment (sindresorhus) [inline renderer/containers/cropper.js:382]: Use early-return.

File: renderer/containers/cropper.js [hunk 3]
```diff
@@ -334,6 +356,50 @@ export default class CropperContainer extends Container {
     }
   };
 
+  redo = () => {
+    if (this.state.undone && this.state.cropperPointer < this.state.cropperHistory.length - 1) {
+      const originalCropperHistory = this.state.cropperHistory;
+      const currPointer = this.state.cropperPointer + 1;
```
Comment (sindresorhus) [inline renderer/containers/cropper.js:362]: Don't use abbreviations.

File: renderer/containers/cropper.js [hunk 4]
```diff
@@ -64,6 +64,10 @@ export default class CropperContainer extends Container {
       isActive: false,
       isReady: false,
       ratio: [1, 1],
+      // Added this to keep track of cropper history to add undo button
+      cropperHistory: [],
+      cropperPointer: -1,
+      undone: false,
```
Comment (sindresorhus) [inline renderer/containers/cropper.js:70]: `isUndone`

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
Would have been nice if the all the undo-related logic would go into a `UndoManager` class.

**Comment by skllcrn** [comment]
Thank you! Great first commit. I'm curious about you reasoning for going with <kbd>Z<kbd> and <kbd>Y<kbd> over <kbd>Command<kbd> + <kbd>Z<kbd> to undo, <kbd>Command<kbd> + <kbd>Shift<kbd> + <kbd>Z<kbd> to redo?

**Comment by lescobar723** [comment]
Hello, thank you for the response! I will try to make the new changes, but I am new to using Electron and React so it will take me a bit but I am up for the challenge! I used the 'z' and 'y' keys for undo and redo because I noticed in the function handleKeyEvent function in the cropper.js file in the pages directory whenever you press one of the keys that are in either in Command+Z and Command+Z+Shift, it really depends on which of those keys you pressed first. Let's say you pressed Shift first in the Command+Z+Shift combination then in the cropper.js file it will go into the Shift case regardless on if you pressed more keys. You cannot really tell the order in which the user will press keys in so I was thinking maybe for multiple key presses the single key cases in handleKeyEvent is not very suitable. So I will try to find a fix to that, one solution is mentioned [here](https://stackoverflow.com/questions/5203407/how-to-detect-if-multiple-keys-are-pressed-at-once-using-javascript). 

Any other suggestions would be greatly appreciated.

File: renderer/pages/cropper.js [hunk 5]
```diff
@@ -20,6 +20,8 @@ actionBarContainer.bindCursor(cursorContainer);
 actionBarContainer.bindCropper(cropperContainer);
 
 let lastRatioLockState = null;
+let metaKeyDown = false;
```
Comment (sindresorhus) [inline renderer/pages/cropper.js:23]:
```suggestion
let iaMetaKeyDown = false;
```

File: renderer/pages/cropper.js [hunk 6]
```diff
@@ -20,6 +20,8 @@ actionBarContainer.bindCursor(cursorContainer);
...
+let ShiftDown = false;
```
Comment (sindresorhus) [inline renderer/pages/cropper.js:24]:
```suggestion
let iaShiftKeyDown = false;
```

File: renderer/pages/cropper.js [hunk 7]
```diff
@@ -96,6 +110,14 @@ export default class CropperPage extends React.Component {
       case 'i':
         this.remote.getCurrentWindow().setIgnoreMouseEvents(true);
         this.dev = !this.dev;
+        break;
+      case 'z':
+        if (metaKeyDown && ShiftDown) {
```
Comment (sindresorhus) [inline renderer/pages/cropper.js:115]: https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/shiftKey

File: renderer/containers/cropper.js [hunk 8]
```diff
@@ -330,7 +356,56 @@ export default class CropperContainer extends Container {
...
+        y,
+        width,
+        height,
+        ratio,
+        displayId
+      });
+    }
+  };
+
+  undo = () => {
```
Comment (sindresorhus) [inline renderer/containers/cropper.js:388]: There's a lot of duplicated logic between the `undo` and `redo` methods. I would try to unify the logic into an internal method which both of these call.

File: renderer/containers/cropper.js [hunk 9]
```diff
@@ -122,8 +126,24 @@ export default class CropperContainer extends Container {
   };
 
   updateSettings = updates => {
-    const {x, y, width, height, ratio, displayId} = this.state;
+    // If already undone, if moved again then erase everything after current pointer
+    if (this.state.isUndone) {
+      let newCropperHistory = this.state.cropperHistory;
+      const CropperHistoryLength = this.state.cropperHistory.length;
```
Comment (sindresorhus) [inline renderer/containers/cropper.js:132]:
```suggestion
      const cropperHistoryLength = this.state.cropperHistory.length;
```

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
Have you looked into using something like https://github.com/frontendphil/react-undo-redo or https://github.com/homerchen19/use-undo ?

**Comment by sindresorhus** [comment]
Closing for lack of response.

**Pull request closed without merging**